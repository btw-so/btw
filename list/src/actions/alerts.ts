import React from 'react';
import { uuid } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export const hideAlert = createAction<string>(ActionTypes.HIDE_ALERT);

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return actionPayload({
      id: options.id || uuid(),
      icon: options.icon || 'dot-circle-o',
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    });
  },
);
