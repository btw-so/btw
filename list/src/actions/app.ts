import { createAction } from '@reduxjs/toolkit';

import { ActionTypes } from 'literals';

import { SetAppOOptions } from 'types';

export const setAppOptions = createAction<SetAppOOptions>(ActionTypes.SET_APP_OPTIONS);
