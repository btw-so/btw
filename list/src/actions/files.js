import { createAction } from "@reduxjs/toolkit";

import { ActionTypes } from "literals";


export const addFile = createAction(
  ActionTypes.ADD_FILE,
  (data) => ({
    payload: data,
  })
);
export const addFileSuccess = createAction(
  ActionTypes.ADD_FILE_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const addFileFailure = createAction(
  ActionTypes.ADD_FILE_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const getFile = createAction(ActionTypes.GET_FILE, (data) => ({
  payload: data,
}));
export const getFileSuccess = createAction(
  ActionTypes.GET_FILE_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const getFileFailure = createAction(
  ActionTypes.GET_FILE_FAILURE,
  (data) => ({
    payload: data,
  })
);
