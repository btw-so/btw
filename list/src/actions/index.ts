export * from './alerts';
export * from './app';
export * from './github';
export * from './user';
export * from './list';
export * from './notes';