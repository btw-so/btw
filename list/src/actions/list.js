import { createAction } from "@reduxjs/toolkit";

import { ActionTypes } from "literals";

export const batchPushNodes = createAction(
  ActionTypes.BATCH_PUSH_NODES,
  (data) => ({
    payload: data,
  })
);

export const getList = createAction(ActionTypes.GET_LIST, (data) => ({
  payload: data,
}));
export const getListSuccess = createAction(
  ActionTypes.GET_LIST_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const getListFailure = createAction(
  ActionTypes.GET_LIST_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const newListNode = createAction(ActionTypes.NEW_LIST_NODE, (data) => ({
  payload: data,
}));

export const editListNodePos = createAction(
  ActionTypes.EDIT_LIST_NODE_POS,
  (data) => ({
    payload: data,
  })
);

export const editListNodeContent = createAction(
  ActionTypes.EDIT_LIST_NODE_CONTENT,
  (data) => ({
    payload: data,
  })
);

export const upsertListNode = createAction(
  ActionTypes.UPSERT_LIST_NODE,
  (data) => ({
    payload: data,
  })
);

export const changeSelectedNode = createAction(
  ActionTypes.CHANGE_SELECTED_NODE,
  (data) => ({
    payload: data,
  })
);

export const getPinnedNodes = createAction(
  ActionTypes.GET_PINNED_NODES,
  (data) => ({
    payload: data,
  })
);

export const getPinnedNodesSuccess = createAction(
  ActionTypes.GET_PINNED_NODES_SUCCESS,
  (data) => ({
    payload: data,
  })
);

export const getPinnedNodesFailure = createAction(
  ActionTypes.GET_PINNED_NODES_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const getPublicNote = createAction(
  ActionTypes.GET_PUBLIC_NOTE,
  (data) => ({
    payload: data,
  })
);

export const getPublicNoteSuccess = createAction(
  ActionTypes.GET_PUBLIC_NOTE_SUCCESS,
  (data) => ({
    payload: data,
  })
);

export const getPublicNoteFailure = createAction(
  ActionTypes.GET_PUBLIC_NOTE_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const searchNodes = createAction(
  ActionTypes.SEARCH_NODES,
  (data) => ({
    payload: data,
  })
);

export const searchNodesSuccess = createAction(
  ActionTypes.SEARCH_NODES_SUCCESS,
  (data) => ({
    payload: data,
  })
);

export const searchNodesFailure = createAction(
  ActionTypes.SEARCH_NODES_FAILURE,
  (data) => ({
    payload: data,
  })
);
