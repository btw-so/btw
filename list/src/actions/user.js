import { createAction } from "@reduxjs/toolkit";

import { ActionTypes } from "literals";

export const login = createAction(ActionTypes.USER_LOGIN_REQUEST);
export const loginSuccess = createAction(ActionTypes.USER_LOGIN_SUCCESS);

export const logOut = createAction(ActionTypes.USER_LOGOUT_REQUEST);
export const logOutSuccess = createAction(ActionTypes.USER_LOGOUT_SUCCESS);

export const getUser = createAction(ActionTypes.GET_USER);
export const getUserSuccess = createAction(
  ActionTypes.GET_USER_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const getUserFailure = createAction(
  ActionTypes.GET_USER_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const generateOtp = createAction(ActionTypes.GENERATE_OTP, (data) => ({
  payload: data,
}));
export const generateOtpSuccess = createAction(
  ActionTypes.GENERATE_OTP_SUCCESS,
  ({ success, error, data }) => ({ payload: { success, error, data } })
);
export const generateOtpFailure = createAction(
  ActionTypes.GENERATE_OTP_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const verifyOtp = createAction(
  ActionTypes.VERIFY_OTP,
  ({ email, otp } = {}) => ({
    payload: { email, otp },
  })
);
export const verifyOtpSuccess = createAction(
  ActionTypes.VERIFY_OTP_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const verifyOtpFailure = createAction(
  ActionTypes.VERIFY_OTP_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const updateUser = createAction(
  ActionTypes.UPDATE_USER,
  ({ name, slug, bio, pic, twitter, linkedin, instagram, settings } = {}) => ({
    payload: { name, slug, bio, pic, twitter, linkedin, instagram, settings },
  })
);
export const updateUserSuccess = createAction(
  ActionTypes.UPDATE_USER_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const updateUserFailure = createAction(
  ActionTypes.UPDATE_USER_FAILURE,
  (data) => ({
    payload: data,
  })
);
export const resetState = createAction(
  ActionTypes.RESET_USER_STATE,
  (data) => ({
    payload: data,
  })
);
