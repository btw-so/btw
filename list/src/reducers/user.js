import { createReducer } from "@reduxjs/toolkit";

import { STATUS } from "literals";

import {
  login,
  loginSuccess,
  logOut,
  logOutSuccess,
  getUser,
  getUserSuccess,
  getUserFailure,
  generateOtp,
  generateOtpSuccess,
  generateOtpFailure,
  verifyOtp,
  verifyOtpSuccess,
  verifyOtpFailure,
  updateUser,
  updateUserSuccess,
  updateUserFailure,
  resetState,
} from "actions";

export const userState = {
  user: {
    status: STATUS.IDLE,
    data: null,
    error: null,
    isLoggedIn: false,
  },
};

export const otpState = {
  otp: {
    status: STATUS.IDLE,
    data: null,
    error: null,
  },
  verifyOtp: {
    status: STATUS.IDLE,
    data: null,
    error: null,
  },
};

export default {
  user: createReducer(userState, (builder) => {
    builder.addCase(resetState, (draft) => {
      draft.user = userState.user;
    });

    builder
      .addCase(getUser, (draft) => {
        draft.user.status = STATUS.RUNNING;
        draft.user.error = null;
      })
      .addCase(getUserSuccess, (draft, { payload }) => {
        draft.user.status = STATUS.SUCCESS;
        draft.user.data = payload;
        draft.user.isLoggedIn = true;
      })
      .addCase(getUserFailure, (draft, { payload }) => {
        draft.user.status = STATUS.ERROR;
        draft.user.error = payload;
        draft.user.isLoggedIn = false;
      });

    builder
      .addCase(updateUser, (draft) => {
        draft.user.status = STATUS.RUNNING;
        draft.user.error = null;
      })
      .addCase(updateUserSuccess, (draft, { payload }) => {
        draft.user.status = STATUS.SUCCESS;
      })
      .addCase(updateUserFailure, (draft, { payload }) => {
        draft.user.status = STATUS.ERROR;
        draft.user.error = payload.error;
      });
  }),

  otp: createReducer(otpState, (builder) => {
    builder.addCase(resetState, (draft) => {
      return otpState;
    });

    builder
      .addCase(generateOtp, (draft) => {
        draft.otp.status = STATUS.RUNNING;
        draft.otp.error = null;
      })
      .addCase(generateOtpSuccess, (draft, { payload }) => {
        draft.otp.status = STATUS.SUCCESS;
      })
      .addCase(generateOtpFailure, (draft, { payload }) => {
        draft.otp.status = STATUS.ERROR;
        draft.otp.error = payload.error;
      });

    builder
      .addCase(verifyOtp, (draft) => {
        draft.verifyOtp.status = STATUS.RUNNING;
        draft.verifyOtp.error = null;
      })
      .addCase(verifyOtpSuccess, (draft, { payload }) => {
        draft.verifyOtp.status = STATUS.SUCCESS;
        draft.verifyOtp.error = null;
      })
      .addCase(verifyOtpFailure, (draft, { payload }) => {
        draft.verifyOtp.status = STATUS.ERROR;
        draft.verifyOtp.error = payload.error;
      });

  }),
};
