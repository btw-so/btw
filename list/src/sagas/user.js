import {
  all,
  delay,
  put,
  takeLatest,
  call,
  takeEvery,
} from "redux-saga/effects";

import toast from "react-hot-toast";

import { ActionTypes } from "literals";

import {
  getUserSuccess,
  getUserFailure,
  generateOtpSuccess,
  generateOtpFailure,
  verifyOtpSuccess,
  verifyOtpFailure,
  updateUserSuccess,
  updateUserFailure,
  resetState,
} from "actions";

import axios from "axios";
const axiosInstance = axios.create({
  timeout: 20000,
  withCredentials: true,
});

import genFingerprint from "../fingerprint";

// import FingerprintJS from '@fingerprintjs/fingerprintjs';

async function getFingerPrint() {
  return genFingerprint();
  // const fp = await FingerprintJS.load();

  // const { visitorId } = await fp.get();

  // return visitorId;
}

export function* getUserSaga() {
  // do a fetch call to process.env.REACT_APP_TASKS_PUBLIC_URL
  // const response = yield call(fetch, 'https://jsonplaceholder.typicode.com/todos/1');
  // const data = yield response.json();

  const fingerprint = yield call(getFingerPrint);

  try {
    const { data: res } = yield call(() =>
      axiosInstance.request({
        url: `${process.env.REACT_APP_TASKS_PUBLIC_URL}/user/details`,
        method: "POST",
        data: {
          fingerprint,
        },
      })
    );

    const { success, data, error } = res;

    if (success && data.user && data.isLoggedIn) {
      yield put(getUserSuccess(data.user));
    } else {
      yield put(getUserFailure({ error: error || "Something went wrong" }));

      if (!data.isLoggedIn) {
        // reset the state
        yield put(resetState());
        // if the user-details API fails, we need to clear the cookie
        // so that the user can login again
        document.cookie = `${
          process.env.REACT_APP_BTW_UUID_KEY || "btw_uuid"
        }=; expires=Thu, 01 Jan 1970 00:00:00 UTC; path=/;`;
        // update the user state fresh
        yield put(getUserFailure({ error: error || "Something went wrong" }));
      }
    }
  } catch (e) {
    yield put(getUserFailure({ error: e.message }));
  }
}

export function* generateOtp({ payload }) {
  const fingerprint = yield call(getFingerPrint);
  const { email } = payload || {};

  try {
    const { data } = yield call(() =>
      axiosInstance.request({
        url: `${process.env.REACT_APP_TASKS_PUBLIC_URL}/otp/generate`,
        method: "POST",
        data: {
          email,
          fingerprint,
        },
      })
    );

    const { success, error } = data;

    if (success) {
      yield put(generateOtpSuccess({ success, error }));
    } else {
      yield put(generateOtpFailure({ error: error || "Something went wrong" }));
    }
  } catch (e) {
    yield put(
      generateOtpFailure({ error: e.message || "Something went wrong" })
    );
  }
}

export function* verifyOtp({ payload }) {
  const fingerprint = yield call(getFingerPrint);
  const { email, otp } = payload || {};

  try {
    const { data: res } = yield call(() =>
      axiosInstance.request({
        url: `${process.env.REACT_APP_TASKS_PUBLIC_URL}/otp/validate`,
        method: "POST",
        data: {
          email,
          otp,
          fingerprint,
        },
      })
    );

    const { success, data, error } = res;

    if (success && data.isValid) {
      // 200ms delay
      yield delay(200);

      // get user details
      yield call(getUserSaga);

      yield delay(200);

      yield put(verifyOtpSuccess());
    } else {
      yield put(verifyOtpFailure({ error: error || "Something went wrong" }));
    }
  } catch (e) {
    yield put(verifyOtpFailure({ error: e.message || "Something went wrong" }));
  }
}

export function* updateUser({ payload }) {
  const fingerprint = yield call(getFingerPrint);
  const { name, slug, bio, pic, twitter, linkedin, instagram, settings } =
    payload || {};

  const toastId = toast.loading("Updating user details");

  try {
    const { data: res } = yield call(() =>
      axiosInstance.request({
        url: `${process.env.REACT_APP_TASKS_PUBLIC_URL}/user/update`,
        method: "POST",
        data: {
          fingerprint,
          name,
          slug,
          bio,
          pic,
          twitter,
          linkedin,
          instagram,
          settings,
        },
      })
    );

    const { success, error } = res;

    if (success) {
      yield put(updateUserSuccess());

      // call getUserSaga to update the user in the store
      yield call(getUserSaga);

      toast.success("Profile updated", {
        id: toastId,
      });
    } else {
      yield put(updateUserFailure({ error: error || "Something went wrong" }));

      toast.error(`Error: ${error}`, {
        id: toastId,
      });
    }
  } catch (e) {
    yield put(
      updateUserFailure({ error: e.message || "Something went wrong" })
    );

    toast.error(`Error: ${e.message}`, {
      id: toastId,
    });
  }
}


export default function* root() {
  yield all([
    takeEvery(ActionTypes.GET_USER, getUserSaga),
    takeEvery(ActionTypes.GENERATE_OTP, generateOtp),
    takeEvery(ActionTypes.VERIFY_OTP, verifyOtp),
    takeEvery(ActionTypes.UPDATE_USER, updateUser),
  ]);
}
