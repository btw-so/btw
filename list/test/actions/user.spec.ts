import { login, logOut } from 'actions/user';

describe('actions/user', () => {
  it('login', () => {
    expect(login()).toMatchSnapshot();
  });

  it('logOut', () => {
    expect(logOut()).toMatchSnapshot();
  });
});
