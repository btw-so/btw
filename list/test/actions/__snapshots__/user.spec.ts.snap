// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions/user logOut 1`] = `
Object {
  "payload": undefined,
  "type": "USER_LOGOUT_REQUEST",
}
`;

exports[`actions/user login 1`] = `
Object {
  "payload": undefined,
  "type": "USER_LOGIN_REQUEST",
}
`;
